### DISCUSSION TITLE
LibGD not converting colour spaces correctly - sRGB Linear to sRGB IEC61966-2.1

### DISCUSSION THREAD

**Discussion by headtailgrep** [libgd/libgd#825] [Ideas]
Friends, I run a large photo sharing wordpress website (large by my standards, but small by most others) that has experienced issues where images, starting as JPG's uploaded by our users, in very very specific circumstances are displayed much darker than they should be. I've traced the issue to the GD library which is used by Wordpress.

To be fair 95% of photos uploaded do not have this issue because the colour space conversion, if necessary, is correct. But the 5% is coming from brand new cameras using exotic colour spaces. For example:

When an image is uploaded and re-sized, it will convert the colourspace if necessary - in this case the JPG was converted by GD from 'SRGB Linear' to 'sRGB IEC61966-2.1' but it does not convert the image properly and it comes out much darker than normal.

Either the conversion is incorrect or the colour space data is not being preserved.

Earlier last year I had a similar issue  with a photographer submitting his work as JPG in the 'ProPhoto RGB' colourspace - unlike SRGB Linear the images would look like white balance was way off. It is basically the same problem.

Of course, converting the colour space in a photoshop to a known SRGB space does solve the problem for the users (if they are willing to perform the conversion - most aren't) but most folks don't know the difference and what's happening is more users are experiencing this as camera manufactures switch to more exotic colour spaces. I now have two users who are now consistently having this issue and it will grow as more photographers purchase newer cameras.

My friend also noticed Flickr was doing similar things to his images. I suspect they may too use GD.

In Summary 'ProPhoto RGB' to 'sRGB IEC61966-2.1' is not correct , as well 'sRGB Linear' to 'sRGB IEC61966-2.1' is also incorrect. Consider preserving colour space in conversions to ensure client image rendering engines can interpret as they need to. Or perhaps it's an issue of colour space interpretation.

There is a bug report filed in 2015 about preserving colourspace - I am by no means an expert in this but it seems as new cameras are coming out this may be an issue we cannot leave behind for much longer.

Thank you all for your time and for this wonderful and most useful code. It's very much appreciated.

**Comment by cmb69** [comment]
This is the same issue as https://github.com/libgd/libgd/issues/136.

**Reply by headtailgrep** [reply]
It is the same or is it similar? #136 says that ICC colour profiles are removed. They aren't removed, they are changed.

I've validated that a re-sized image (originally uploaded with ICC profile SRGB Linear) by GD on my site replaces the original ICC profile with "sRGB IEC61966-2.1'

So what may be happening is it isn't properly re-interpreting the image before changing the ICC profile. Perhaps this is a bug in the re-interpretation or perhaps it's not interpreted at all (it's ignored) and the name of the profile is simply changed in error?

Meaning a correct resolution to this issue is to preserve the original colour space/ICC profile in the GD output and let the browser do the interpretation of colour as per the profile.

**Reply by cmb69** [reply]
libgd currently completely ignores color profiles when loading an image. When the image is written back, some default color profile might be added.

**Reply by pierrejoye** [reply]
We do not add any profile. The respective library may add one by default, however GD does not nothing but using the pure raw RGB(A) data and store it again.

We do plan to use littleCMS in 3.5 or maybe 3.4, first to keep the profiles (and other informations), then to allow to expose littleCMS resources so one could process them as well.

**Comment by pierrejoye** [comment]
@headtailgrep could you provide some details about the input and output formats? Or if you could send the image?

**Reply by headtailgrep** [reply]
Sorry for the delay, and thank you for your time.

I've created a package with two test images: One image is the original in "SRGB - Linear" colour space. The other is a re-sized thumbnail processed by GD with the 'new' colourspace. I've also included screenshots from photoshop from each image that reports what colour space these images are in. And a text file with these details. Download from my server here:

http://steve.hostovsky.com/imagetest_gd.zip and is only 1.8 megs in size

If you don't trust my server, check the root and see i'm just a hacker/photographer hehehe :)

I can send by some other method if you want.. let me know and thank you again everyone. Once enough folks have a copy I'll remove it and this link.

**Reply by pierrejoye** [reply]
Long ago, but for the record in this conversion :)

sRGB format, set on export is easy (same as converting to sRGB using another tool). We could set it export. For such options, and other per codec, new APIs are on the work to allow more custom options to be set on read/write, to give full control to the caller of what the format should be using the underlying library, if any, available options.

The challenge is actually all ops that could be applied to that image before saving. The normal flow would be:

decode sRGB > linearize  > ops > re-encode to sRGB

now, without that linearize step and re-encoding, it will remain gamma wrong and ops won't be applied in linear light, dark halo or releated artififacts like with resampling etc.
